### DISCUSSION TITLE
How to use MemoryDatabase? (for testing purpose)

### DISCUSSION THREAD

**Discussion by Odonno** [cyrilletuzi/angular-async-local-storage#543] [General]
# Checks before posting an issue

- [x] I have read all the library's documentation (README, CHANGELOG, MIGRATION and /docs)
- [x] I understand I am writing to a *human being*, doing this open source project on his *free* time
- [x] I will be involved in my own issue, by answering questions and informing if the issue is resolved

# Configuration

- `@ngx-pwa/local-storage` version: 11.1.0
- Angular version (`ng version`):
```
Angular CLI: 11.1.1
Node: 12.18.3
OS: win32 x64

Angular: 11.1.0
... animations, common, compiler, compiler-cli, core, forms
... language-service, localize, platform-browser
... platform-browser-dynamic, router, service-worker
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1100.7
@angular-devkit/build-angular   0.1100.7
@angular-devkit/core            11.1.1
@angular-devkit/schematics      11.1.1
@angular/cdk                    11.0.3
@angular/cli                    11.1.1
@angular/fire                   6.1.4
@angular/material               11.0.3
@schematics/angular             11.1.1
@schematics/update              0.1101.1
rxjs                            6.6.3
typescript                      4.0.5
```

# Description of the issue

How to use MemoryDatabase? (for testing purpose)

I would like to use an in-memory database to be able to write unit tests in my Angular app.

**Comment by cyrilletuzi** [comment]
Hi,

The `MemoryDatabase` is an _automatic_ fallback when `indexedDb` or `localStorage` are not available.

So either `indexedDb` or `localStorage` is available in the context of your unit tests (which should be the case as unit tests are run in browsers), so there should be no problem, or `MemoryDatabase` will be used automatically.

Why do you want to disable the normal behavior of the lib?

**Reply by Odonno** [reply]
I would like to inject a `MemoryStorage` inside the `TestBed` so I can control what's inside the storage. You could provide something like the `HttpClientTestingModule` using the same pattern. What do you think?

Like providing a `StorageTestingModule`.

**Comment by cyrilletuzi** [comment]
> Like providing a StorageTestingModule.

It could be a possibility.

Another possibility would be to export the different databases type (so `MemoryDatabase` among them) as public, so you would be able to override `LocalDabatabase` via `NgModule` `providers` (note that you can already do so with your own implementation, or by copying the [`MemoryDatabase` class](https://github.com/cyrilletuzi/angular-async-local-storage/blob/main/projects/ngx-pwa/local-storage/src/lib/databases/memory-database.ts)).

But first I would like to be sure to understand your goal / use case, because it's not clear to me yet.

> "control what's inside the storage"

Why not control the real storage directly?

> You could provide something like the `HttpClientTestingModule` using the same pattern.

The `HttpClientTestingModule` makes sense because during unit tests, the API is not accessible. But the storage is.

**Reply by Odonno** [reply]
Ok, nevermind. I managed to make Jest functions and it is enough for my needs.
